### PR TITLE
[DEV-2515] Add mongo migration for feature entity relationship & primary entity derivation

### PR DESCRIPTION
**Pull request by gee-senbong**
## Description

Add migration script for `relationships_info` & `primary_entity_ids` attributes. These attributes are introduced in 
* https://github.com/featurebyte/featurebyte/pull/1874 (`relationships_info`)
* https://github.com/featurebyte/featurebyte/pull/1866 (`primary_entity_ids`)

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/docker/featurebyte.yml [hunk 1]
```diff
@@ -12,6 +12,9 @@ services:
       - MONGO_HOSTNAME=mongo-rs
     volumes:
       - mongodb:/data/
+    ports:
+      - "0.0.0.0:27021:27021"
+      - "0.0.0.0:27022:27022"
```
Comment (gee-senbong) [inline featurebyte/docker/featurebyte.yml:17]: Expose these 2 ports for the ease of debugging.

File: featurebyte/migration/service/feature.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+                relationship.dict(by_alias=True) for relationship in relationships_info
+            ]
+        return documents
+
+    def migrate_document_record(self, record: Dict[str, Any]) -> Dict[str, Any]:
+        try:
+            output = super().migrate_document_record(record)
+            return output
+        except Exception as exc:  # pylint: disable=broad-except
+            # for some bad records, we just log the error and return the original record
```
Comment (jevonyeoh) [inline featurebyte/migration/service/feature.py:98]: curious - do we know what the common causes for failure are?

File: featurebyte/migration/service/feature.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+                "Record (id: %s, name: %s) encountered error: %s",
+                record["_id"],
+                record["name"],
+                str(exc),
+            )
+            return record
+
+    @migrate(version=4, description="Add relationships_info to feature record")
+    async def add_relationships_info(self) -> None:
+        """Change field name from event_data_ids to tabular_data_ids"""
```
Comment (jevonyeoh) [inline featurebyte/migration/service/feature.py:110]: do we need to update this docstring/

File: featurebyte/migration/service/feature.py (hunk 2 above)
Comment (gee-senbong) [reply featurebyte/migration/service/feature.py:98]: For the older records where `aggregation_ids` is missing (
https://github.com/featurebyte/featurebyte/blob/2aef0b53d0a51e23645dd275c8ff500743bc971d/featurebyte/models/feature.py#L452), the root validator attempts to derive the attribute. It fails to generate the attribute when deriving operation structure of the join node.

File: featurebyte/migration/service/feature.py (hunk 3 above)
Comment (gee-senbong) [reply featurebyte/migration/service/feature.py:110]: Good catch, will update the docstring here.

File: featurebyte/routes/registry.py [hunk 4]
```diff
@@ -337,6 +336,11 @@
 app_container_config.register_class(BlockModificationHandler)
 app_container_config.register_class(MongoDBImpl, name_override="persistent")
 
+# register migration services
+app_container_config.register_class(DataWarehouseMigrationServiceV1)
+app_container_config.register_class(DataWarehouseMigrationServiceV3)
```
[the remark this answers is not in the document]
Comment (gee-senbong) [reply featurebyte/routes/registry.py:341]: `DataWarehouseMigrationServiceV1` contains migration version 1 & version 2. 

https://github.com/featurebyte/featurebyte/blob/main/featurebyte/migration/service/data_warehouse.py#L148
https://github.com/featurebyte/featurebyte/blob/main/featurebyte/migration/service/data_warehouse.py#L155-L159

[the remark this answers is not in the document]
Comment (jevonyeoh) [reply featurebyte/routes/registry.py:341]: ah yes, always forget that that's where the v2 is

File: featurebyte/migration/service/feature.py (hunk 3 above)
Comment (gee-senbong) [reply featurebyte/migration/service/feature.py:110]: Will update the docstring in the coming PR.

**Pull request merged**